htlcswitch/hop/payload.go

```diff
@@ -410,3 +417,85 @@ func getMinRequiredViolation(set tlv.TypeMap) *tlv.Type {
 
 	return nil
 }
+
+// ValidateBlindedRouteData performs the additional validation that is
+// required for payments that rely on data provided in an encrypted blob to
+// be forwarded. We enforce the blinded route's maximum expiry height so that
+// the route "expires" and a malicious party does not have endless opportunity
+// to probe the blinded route and compare it to updated channel policies in
+// the network.
+func ValidateBlindedRouteData(blindedData *record.BlindedRouteData,
+	incomingAmount lnwire.MilliSatoshi, incomingTimelock uint32) error {
+
+	// Bolt 04 notes that we should enforce payment constraints _if_ they
+	// are present, so we do not fail if not provided.
+	if blindedData.Constraints != nil {
+		// MUST fail if the expiry is greater than max_cltv_expiry.
+		maxCLTV := blindedData.Constraints.MaxCltvExpiry
+		if incomingTimelock > maxCLTV {
+			return ErrInvalidPayload{
+				Type:      record.LockTimeOnionType,
+				Violation: InsufficientViolation,
+			}
+		}
+
+		// MUST fail if the amount is below htlc_minimum_msat.
+		if incomingAmount < blindedData.Constraints.HtlcMinimumMsat {
+			return ErrInvalidPayload{
+				Type:      record.AmtOnionType,
+				Violation: InsufficientViolation,
+			}
+		}
+	}
+
+	// Encrypted data MUST include payment relay information.
+	if blindedData.RelayInfo == nil {
+		return ErrInvalidPayload{
+			Type:      record.PaymentRelayType,
+			Violation: OmittedViolation,
+		}
+	}
+
+	// Specification validation instructions for receivers are:
+	// * encrypted data must either have short_channel id _or_ next_node_id.
+	//
+	// However, instructions for creators of blinded route encrypted data
+	// for payments state that a short_channel_id MUST be included.
+	//
+	// This difference in validation and creation instructions is present
+	// because blinded routes are used for blinded payments and onion
+	// messages (the latter relying on node IDs rather than short channel
+	// ids).
+	//
+	// While we may transition relaying to node_id based payments, this is
+	// not currently a valid option for blinded payments (because the
+	// receiver must give us a short channel ID), so we simply validate
+	// that we have a short channel ID.
+	//
+	// TODO: relax validation if node_id based blinded payments are
+	// introduced to the specification.
+	if blindedData.ShortChannelID == nil {
+		return ErrInvalidPayload{
+			Type:      record.ShortChannelIDType,
+			Violation: OmittedViolation,
+		}
+	}
+
+	// No need to check anything else if features are not provided (bolt 4
+	// indicates that omitted features should be treated like an empty
+	// vector).
+	if blindedData.Features == nil {
+		return nil
+	}
+
+	// Fail if we don't understand any features (even or odd), because we
+	// expect the features to have been set from our announcement.
+	if blindedData.Features.UnknownFeatures() {
```

Comment:
The creator specs say that if `allowed_features` is set, it must set it to an empty array, but in the forwarding case we are less restrictive, to be able to set some features in the future I guess. The question is then to which features we should compare to, to the node's feature set or to the channel's feature set (there are currently none specified https://github.com/lightning/bolts/blob/3b814f0d031450a9cddd2718be7ed50c5444095b/09-features.md)?

Reply:
>  to be able to set some features in the future I guess.

Yeah this is future proofing us for upgrades to things like PTLCs, where we want to make sure we support the payment type. Since creators can't set the feature, I think we can leave this out of scope (at least until LND is creating its own blinded paths?)

> one could check if tlv onions are advertised and then not forward if encountered

I'll clarify this is the spec, but I _think_ that `tlv_onion` doesn't need to be enforced here as it's implicitly supported by our ability to parse the payload in the first place?